### PR TITLE
Respect array alpha with interpolation_stage='rgba' in _Imagebase::_make_image

### PR DESCRIPTION
**Pull request by chaoyihu**
## PR summary

This PR hopefully closes https://github.com/matplotlib/matplotlib/issues/28382, making sure that array alpha is respected in `_Imagebase::_make_image` when param `interpolation_stage` is set to 'rgba'.

## PR checklist

- [x] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [N/A] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [N/A] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/api_changes.html#announce-changes-deprecations-and-new-features)
- [N/A] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

### CODE REVIEW

**Comment by chaoyihu** [comment]
Hi @timhoffm, may I request your review on this pr?

Sorry for the tagging - I'm not sure how to request a review when there's no suggested reviewers on the GitHub interface.

**Review by jklymak — COMMENTED** [review]
This should be clarified and the tests expanded a bit.  What happens if A has an alpha channel and *alpha* is also a matrix for `interpolation_stage='rgba'`?  We have a behaviour for `interpolation_stage='data'` so this PR should have the same behaviour if possible, and it'll need a test.

**Comment by chaoyihu** [comment]
I tested with all possible combinations, each repeated with `interpolation_stage` as `data` and `rgba`, respectively:
1. When the image does not have alpha channel:
a. alpha not specified: default to 1.0.
b. scalar alpha: broadcast.
c. array alpha: concatenate.
2. When the image has an alpha channel:
a. alpha not specified: nothings needs to be done.
b. scalar alpha: multiply the alpha channel with user-specified scalar.
c. array alpha: replace the alpha channel with user-specified array.

The expected outcomes seem to align with existing test cases. I've observed conflicts when experimenting with different behaviors, for example, `test_image::test_image_composite_alpha` fails when I change 2.b. to replacing instead of multiplying.

I'm assuming the code is fine since it passed all checks, but please let me know if I'm missing anything.

File: lib/matplotlib/image.py [hunk 1]
```diff
@@ -550,20 +550,33 @@ def _make_image(self, A, in_bbox, out_bbox, clip_bbox, magnification=1.0,
                      cbook._setattr_cm(self.norm, vmin=s_vmin, vmax=s_vmax):
                     output = self.norm(resampled_masked)
             else:
-                if A.ndim == 2:  # _interpolation_stage == 'rgba'
+                if A.ndim == 2:  # self._interpolation_stage == 'rgba'
```
Comment (timhoffm) [inline lib/matplotlib/image.py:553]: The comment would better be placed on the `else` above.

File: lib/matplotlib/tests/test_image.py [hunk 2]
```diff
@@ -1593,3 +1593,39 @@ def test_resample_dtypes(dtype, ndim):
...
     axes_image.make_image(None)[0]
+
+
+@pytest.mark.parametrize('intp_stage', ('data', 'rgba'))
+@check_figures_equal()
+def test_interpolation_stage_rgba_respects_alpha_param(fig_test, fig_ref, intp_stage):
+    axs_tst = fig_test.subplots(2, 3)
+    axs_ref = fig_ref.subplots(2, 3)
+    scl_a = 0.5  # scalar alpha
+    arr_a = np.random.rand(3, 3)  # array alpha
```
Comment (timhoffm) [inline lib/matplotlib/tests/test_image.py:1604]: Use a telling variable name instead of a cryptic one, which is explained in a comment 😃. 

To have less magic numbers and keep it more readable, I propose to variables for the number of points.

```suggestion
    ny, nx = 3, 3
    scalar_alpha = 0.5
    array_alpha = np.random.rand(ny, nx)
```

File: lib/matplotlib/tests/test_image.py [hunk 3]
```diff
@@ -1593,3 +1593,39 @@ def test_resample_dtypes(dtype, ndim):
...
+
+    # When the image does not have an alpha channel, alpha should be specified
+    # by the user or default to 1.0
+    im1 = np.random.rand(3, 3, 3)
+    im_concat_arr_a = np.concatenate((im1, np.expand_dims(arr_a, axis=-1)), axis=-1)
```
Comment (timhoffm) [inline lib/matplotlib/tests/test_image.py:1609]: I believe this is more easy to read than concat/expand_dims:
```suggestion
    im_rgb = np.random.rand(ny, nx, 3)
    im_rgba = np.empty(ny, nx, 4)
    im_rgba[..., :3] = im_rgb
    im_rgba[..., 3] = array_alpha
```

File: lib/matplotlib/image.py [hunk 4]
```diff
@@ -550,20 +550,33 @@ def _make_image(self, A, in_bbox, out_bbox, clip_bbox, magnification=1.0,
...
+                elif np.ndim(alpha) > 0:  # Array alpha
+                    # user-specified array alpha overrides the existing alpha channel
+                    output_alpha = _resample(self, alpha, out_shape, t)
+                    output = _resample(
+                        self, _rgb_to_rgba(A[..., :3]), out_shape, t)
+                else:  # Scalar alpha
+                    if A.shape[2] == 3:  # broadcast scalar alpha
+                        output_alpha = (255 * alpha) if A.dtype == np.uint8 else alpha
+                    else:  # or apply scalar alpha to existing alpha channel
+                        output_alpha = _resample(self, A[..., 3], out_shape, t) * alpha
```
Comment (timhoffm) [inline lib/matplotlib/image.py:577]: To be checked, but I think and explicit alpha should overwrite existing alpha rather than scaling it.

File: lib/matplotlib/image.py (hunk 1 above)
Comment (chaoyihu) [reply lib/matplotlib/image.py:553]: I think it should stay with the `if` statement?
The `else` above is the logical negation of `if A.ndim == 2 and self._interpolation_stage == 'data'`, which covers `A.ndim != 2` when `self._interpolation_stage` can be either `'data'` or `'rgba'`.

Comment (timhoffm) [reply lib/matplotlib/image.py:553]: Yes, true. Sorry for the confusion.

File: lib/matplotlib/tests/test_image.py (hunk 3 above)
Comment (chaoyihu) [reply lib/matplotlib/tests/test_image.py:1609]: I prefer to avoid creating variables that are only used once.
How about this?
```
    im_rgb = np.random.rand(ny, nx, 3)
    im_concat_arr_a = np.concatenate(  # combine rgb channels with array alpha
        (im_rgb, array_alpha.reshape((3, 3, 1))), axis=-1
    )
```

File: lib/matplotlib/image.py (hunk 4 above)
Comment (chaoyihu) [reply lib/matplotlib/image.py:577]: Changing this will cause conflicts:

> I've observed conflicts when experimenting with different behaviors, for example, test_image::test_image_composite_alpha fails when I change 2.b. to replacing instead of multiplying.

This test case creates color stripes with fading alpha peaking at 1.0 at the center, and applied scalar alpha to the stripe at different locations, for example:
```
    arr = np.zeros((11, 21, 4))
    arr[:, :, 0] = 1
    arr[:, :, 3] = np.concatenate(
        (np.arange(0, 1.1, 0.1), np.arange(0, 1, 0.1)[::-1]))
    ax.imshow(arr, extent=[1, 2, 5, 0], alpha=0.3)
```
And the baseline image uses a combined alpha, see below. (If it was overwritten by a scalar the fading effects would have been removed.)
image_composite_alpha

File: lib/matplotlib/tests/test_image.py (hunk 3 above)
Comment (chaoyihu) [reply lib/matplotlib/tests/test_image.py]: Sorry, my new commit made this outdated. Do you still prefer using temporary variables? If so I'd be happy to change it.

**Review by jklymak — APPROVED** [review]
This looks right to me.  I think it probably needs a very short change note under `doc/api/next_api_changes/behavior/` before merging.  Thanks for digging into this!

**Comment by jklymak** [comment]
#28061 went in so unfortunately you now have merge clashes, and will need to rebase.  However I think its pretty straight forward, but let us know if you have difficulties - if you havent' done these before, it'd be worth backing up your branch...

File: lib/matplotlib/image.py (hunk 4 above)
Comment (timhoffm) [reply lib/matplotlib/image.py:589]: Ok there is currently a different behavior between scalar and array alpha if the image already has an alpha.

The  [documentation](https://matplotlib.org/stable/api/_as_gen/matplotlib.axes.Axes.imshow.html) is not clear about the intended behavior:

> The alpha blending value, between 0 (transparent) and 1 (opaque). If alpha is an array, the alpha blending values are applied pixel by pixel, and alpha must have the same shape as X.

My point is that this behavior is quite surprising and inconsistent:

- scalar `alpha`: multiply the alpha channel with user-specified scalar.
- array `alpha`: replace the alpha channel with user-specified array.

In all other cases, when we have a rgba color and an explicit alpha, the explicit alpha overwrites the rgba alpha; effectively through https://matplotlib.org/stable/api/_as_gen/matplotlib.colors.to_rgba.html / https://matplotlib.org/stable/api/_as_gen/matplotlib.colors.to_rgba_array.html

Additionally, I would have expected that scalar and array behave the same - either the scalar scales all elements and then the array also scales element-wise, or the both the scalar and the array overwrite.

Since tests cover the current behavior, this PR should be fine in that it does not introduce a behavior change. OTOH I'm wondering whether this behavior is really intended. The test and alpha handling came in in #1955, but this PR only states to implement missing alpha handling. There has been no discussion whether overwriting or multiplying is the right choice.

File: lib/matplotlib/tests/test_image.py (hunk 3 above)
Comment (timhoffm) [reply lib/matplotlib/tests/test_image.py]: No strong opinion. I find it slightly more readable to say "the first three components are `im_rgb`, the fourth component is `array_alpha` rather than "is a concantenation of `im_rgb` and `array_alpha` along the most inner axis.

Note that, you can write `array_alpha.reshape((3, 3, 1)` as `array_alpha[..., np.newaxis]`, which IMHO makes the intent a bit more clear. also `concatenate(..., axis=-1)` is equivalent to `dstack(...)` in this case. So
```
# combine rgb channels with array alpha
im_concat_arr_a = np.dstack((im_rgb, array_alpha[..., np.newaxis]))  
```
would also be an option.

I'll leave it to you to decide which way you want to go.

File: lib/matplotlib/image.py (hunk 4 above)
Comment (chaoyihu) [reply lib/matplotlib/image.py:589]: > My point is that this behavior is quite surprising and inconsistent

> The [documentation](https://matplotlib.org/stable/api/_as_gen/matplotlib.axes.Axes.imshow.html) is not clear about the intended behavior

It should be possible to make the behaviors consistent. Forcing scalar alpha to overwrite causes conflicts, while multiplying array alpha pixel-wise does not seem to break any existing test case in `test_image.py`. Not sure if it has any side effects since I haven't ran the full test suite. Do you want me to try to change it this way?

This is a design choice so it's up to you. Either way, I can clarify the behavior in the docstring of `imshow`.

> when we have a rgba color and an explicit alpha, the explicit alpha overwrites the rgba alpha; effectively through `matplotlib.colors.to_rgba` / `matplotlib.colors.to_rgba_array`

I see two methods of the same name `to_rgba` in the code base:
- [`colors.to_rgba`](https://matplotlib.org/stable/api/_as_gen/matplotlib.colors.to_rgba.html): If alpha is given, force the alpha value of the returned RGBA tuple to alpha.
- [`cm.ScalarMappable.to_rgba`](https://matplotlib.org/stable/api/cm_api.html#matplotlib.cm.ScalarMappable.to_rgba): If the last dimension is 4, the alpha kwarg is ignored; it does not replace the preexisting alpha.

The second one is directly invoked in `image.py` thus looks more relevant to me. It shows yet another different behavior, but I think it is specific to this method for utility purposes and does not imply any design consideration about alpha handling.

I'm not sure about `to_rgba_array`, but intuitively I think it can be used as a setter and not for visualization purposes, so replacing preexisting alpha makes sense in its case.

> The test and alpha handling came in in https://github.com/matplotlib/matplotlib/pull/1955, but this PR only states to implement missing alpha handling.

I think this PR and its related test case should be deemed correct - intended or not, because it does not cause conflicts anyway, either if we choose to keep the current behavior, or multiply array alpha.

Comment (timhoffm) [reply lib/matplotlib/image.py:589]: @jklymak @anntzer I would like your opinion as image/interpolation experts.

The underlying problem here is that for RGB and RGBA images the kwarg *alpha* has been ignored if it was an array. The solution for RGB is straight forward (just use the given alpha array). However, what is the expected behavior for RGBA images, i.e.  `imshow(rgba_array, alpha=alpha)`?

Should the kwarg *alpha* replace the intrinsic alpha channel? Generally I would say yes, this is how alpha also works in other cases (which is coded in [color.to_rgba_array](https://matplotlib.org/stable/api/_as_gen/matplotlib.colors.to_rgba.html)). The problem is that a scalar *alpha* kwarg does not exhibit this behavior, instead `imshow(rgba_array, alpha=0.5)` will scale the intrinsic alpha channel rather than replacing it.

Basically we have three options:
- array alpha replaces intrinsic alpha - we live with the API inconsistency (scalar alpha scales, array alpha replaces)
- array alpha scales intrinsic alpha - this is consistent with the existing scalar alpha for images, but different to the usual behavoir of `alpha` kwargs.
- array alpha and scalar alpha replace intrinsic alpha, and intrinsic - this means an API change for scalar alpha on images.

Comment (jklymak) [reply lib/matplotlib/image.py:589]: Yes, I guess I could go either way, but I think for simplicity alpha should either _replace_ or _scale_ regardless of whether it is scalar or an array.

I guess the case for _scale_ is "I have an image (that has some alpha-d colors) and now I want to partially mask certain regions even more", and that is a bit more complicated and the use most folks would already have for non-alpha channel images.

Conversely, the case for _replace_ is consistency with the rest of the library, where I think _replace_ is what folks expect from *alpha*.   Asking someone to do the math to calculate their own alpha is not too hard.

Comment (timhoffm) [reply lib/matplotlib/image.py:589]: If we can't decide, we could also cheat our way out and only do the operation if all RGBA alpha values are 1 (which I assume is the case for `interpolation_stage='rgba'`) and otherwise either (i) leave this as is (=do nothing) which is the minimal change or (ii) error out and claim that array alpha and RGBA is not supported.

Comment (timhoffm) [reply lib/matplotlib/image.py:589]: I can live with that. Then let's take this PR basically as is so that we can release it as a bugfix in 3.10.1, which is the important part. We can then decide how to change/migrate the current behavior for RGBA and scaler alpha.

Comment (jklymak) [reply lib/matplotlib/image.py:589]: Sounds good to me. But I know @tacaswell and @anntzer have thoughts on this as well.

**Comment by ksunden** [comment]
We are looking to include this in the 3.10.1 release, so would like to move forward.

I think the only outstanding comments on this are rewriting the change note and a question as to whether an internal function requires a parameter after this change. My leaning is to accept the rewritten change note (with or without the additional comment in the proposed change above) and not touch the internal function at this time.

I intend to merge the proposed change note rewrite in approximately 20 hours unless there is a) dissent expressed or b) someone else beats me to it. (I'd have said 24, but would like to get the release going earlier in my work day if I'm doing it tomorrow...)

**Comment by ksunden** [comment]
I'm noting that there was a failed windows test, for which the logs have expired. I'm about 95% sure that is an unrelated flaky test or two, but the tests will be rerun on merging the rewritten change note anyway, so not going to worry about it too much until then.

File: lib/matplotlib/image.py (hunk 4 above)
Comment (tacaswell) [reply lib/matplotlib/image.py:589]: I agree taking this as is.

I however disagree that multiplying in the scalar case for already RGBA data is inconsistent.  In the case of lines the layer we add when rendering has variable alpha if anti-aliasing is turned on.  When we set the alpha on a line we expect the result to be the combination of the AA alpha + the user alpha.  

By analogy, if the user provides us an RGBA image with non-unifrom alpha _and_ provides an artist-level alpha the expectation would be that they are combined not replaced.

**Comment by tacaswell** [comment]
I took the liberty of rebasing to fix the docs build (circle CI runs on the branch tip (and gets its own config from the branch tip) where as GHA runs on a "synthetic" merge of the branch into the target branch.  To make all our CI look "the same" we do the merge as part of the CI workflow, however this means when we bump the minimum python past what we are using on circle you we get "can not install Matplotlib" errors).

File: lib/matplotlib/image.py (hunk 4 above)
Comment (anntzer) [reply lib/matplotlib/image.py:522]: From a consistency PoV I have a mild preference for scaling, which at least is internally consistent within imshow() (e.g. if we decide to also handle e.g. alpha=[[scalar]], a 1x1 array which is broadcastable to a 2D array, our life will be easier if we choose to scale) even though it is not consistent with *other* functions of matplotlib (but that's a secondary concern IMO).
But I can't say that solution is really great either, so I don't feel very strongly either way.

**Comment by anntzer** [comment]
Per the above: mild preference for scaling rather than replacing preexisting alphas with the array alpha (sorry for not commenting earlier), but won't block either way.

File: lib/matplotlib/image.py (hunk 4 above)
Comment (jklymak) [reply lib/matplotlib/image.py:522]: The difference here, versus an antialiased line, is that the user has access to the alpha channel.  I still think simpler is better.  If someone sets alpha with a scalar and they meant it to be multiplicative, they can go back to their original alpha array and do the multiplication easily enough.

**Review by tacaswell — APPROVED** [review]

**Review by ksunden — APPROVED** [review]
